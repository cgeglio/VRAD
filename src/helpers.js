export const fetchDetails = (areas) => {
  const promises = areas.areas.map(area => {
    return fetch('http://localhost:3001' + area.details)
      .then(response => {
        if (!response.ok) {
          throw Error('Error gathering listings')
        }
        return response.json()
      })
      .then(areaInfo => {
        return {
          id: areaInfo.id,
          nickname: area.name || area.area,
          longname: areaInfo.name,
          location: areaInfo.location,
          description: areaInfo.about,
          listings: areaInfo.listings,
        };
      })
  });
  return Promise.all(promises)
};

export const getAreas = () => {
  return fetch('http://localhost:3001/api/v1/areas/')
  .then(response => {
      if (!response.ok) {
        throw Error('Error while fetching, did not get 200 status code')
      }
      return response.json()
    });
  };

export const getListings = (listings) => {
  const allListings = listings.map(listing => {
    return fetch(`http://localhost:3001${listing}`)
      .then(response =>  {
        if (!response.ok) {
          throw Error('Error while fetching, did not get 200 status code')
        }
        return response.json()
      });
    });
 return Promise.all(allListings)
};
