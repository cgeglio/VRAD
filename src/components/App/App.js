import React, { Component } from 'react';
import './App.scss';
import ListingContainer from '../ListingContainer/ListingContainer'
import ListingDetailsContainer from '../ListingDetailsContainer/ListingDetailsContainer'
import FavoritesContainer from '../FavoritesContainer/FavoritesContainer'
import { fetchDetails, getAreas } from '../../helpers';
import Loader from '../Loader/Loader'
import Form from '../Form/Form'
import { Header } from '../Header/Header'
import { AreasContainer } from '../AreasContainer/AreasContainer'
import { Route, Redirect, Link } from 'react-router-dom';
import AreaMap from '../AreaMap/AreaMap';

class App extends Component {
  constructor() {
    super();
    this.state = {
      isLoggedIn: false,
      areas: '',
      error: '',
      user: { name: '', email: '', purpose: '' },
      isLoading: true,
      listings: [],
      currentListing: '',
      favorites: [],
    };
  }

  componentDidMount() {
    getAreas()
      .then(areas => fetchDetails(areas))
      .then(areaData => this.setState({ areas: areaData, isLoading: false }))
      .catch(error => this.setState({ error: 'Encountered error' }));
  }

  addUser = (user) => {
    this.setState({ user, isLoggedIn: true });
  }

  logout = () => {
    this.setState({ isLoggedIn: false, user: { name: '', email: '', purpose: '' }, favorites: [] });
  }

  addListingsToState = listings => {
    const allListings = listings.map(listing => listing);
    this.setState({ listings: allListings });
  }

  setCurrentListing = listing => {
    this.setState({ currentListing: listing });
  }

  addFavorite = listing => {
    const { favorites } = this.state;
    const ids = favorites.map(favorite => favorite.listing_id);
    return !ids.includes(listing.listing_id)
      ? this.setState({ favorites: [...favorites, listing] })
      : this.removeFavorite(listing.listing_id);
  }

  removeFavorite = listingId => {
    const { favorites } = this.state;
    this.setState({ favorites: favorites.filter(favorite => favorite.listing_id !== listingId) });
  }

  render() {
    const { favorites, user, listings, isLoading, isLoggedIn, areas, currentListing } = this.state;
    return (
      <main className="app">
        <Route exact path="/">
          <>
            {isLoggedIn ? <Redirect to="/areas" /> : (
              <>
                <img src={process.env.PUBLIC_URL + '/scoutGreen.png'} alt="Logo" className="logo" />
                <img src={process.env.PUBLIC_URL + '/rmnp.jpg'} alt="Rocky Mountain National Park" className="background-img" />
                <Form
                  addUser={this.addUser}
                />
              </>
            )}
          </>
        </Route>
        <Route exact path="/areas" render={() => {
          return !isLoggedIn ? <Redirect to="/" /> : (
            <>
              <Header
                favoritesNumber={favorites.length}
                user={user}
                logout={this.logout}
              />
              {!isLoading
                ? (
                  <>
                    <AreasContainer
                      areas={areas}
                      addListingsToState={this.addListingsToState}
                    />
                    <AreaMap />
                  </>
                )
                : <Loader />}
            </>
          );
        }}
        />
        <Route exact path="/area/:id/listings" render={() => {
          return !isLoggedIn ? <Redirect to="/" /> : (
            <>
              <Header
                favoritesNumber={favorites.length}
                user={user}
                logout={this.logout}
              />
              <ListingContainer
                favorites={favorites}
                listings={listings}
                addFavorite={this.addFavorite}
                setCurrentListing={this.setCurrentListing}
              />
              <Link to="/areas" className="back-btn">Back to Areas</Link>
            </>
          );
        }}
        />
        <Route exact path="/area/:id/listings/:id" render={() => {
          return !isLoggedIn ? <Redirect to="/" /> : (
            <>
              <Header
                favoritesNumber={favorites.length}
                user={user}
                logout={this.logout}
              />
              <ListingDetailsContainer
                {...currentListing}
              />
              <Link to="/area/:id/listings" className="back-btn">Back to Listings</Link>
            </>
          );
        }}
        />
        <Route exact path="/favorites" render={() => {
          return !isLoggedIn ? <Redirect to="/" /> : (
            <>
              <Header
                favoritesNumber={favorites.length}
                user={user}
                logout={this.logout}
              />
              {!favorites.length
                ? (
                  <div className="page">
                    <h2 className="no-favorites">You don't have any favorites yet!</h2>
                    <Link to="/area/:id/listings" className="back-btn">Back to Listings</Link>
                  </div>
                )
                : (
                  <div className="page">
                    <FavoritesContainer
                      favorites={favorites}
                      addFavorite={this.addFavorite}
                      setCurrentListing={this.setCurrentListing}
                    />
                    <Link to="/area/:id/listings" className="back-btn">Back to Listings</Link>
                  </div>
                )}
            </>
          );
        }}
        />
      </main>
    );
  }
}

export default App;
