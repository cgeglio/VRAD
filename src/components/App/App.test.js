import React from 'react';
import { shallow } from 'enzyme';
import App from './App';
import { fetchDetails, getAreas } from '../../helpers.js';

jest.mock('../../helpers.js');

describe('App', () => {
  let wrapper;
  let mockListing;
  let mockUser;
  let mockArea;

  beforeEach(() => {
    mockListing = { name: 'listing 1', listing_id: 5 };
    mockUser = { name: 'Jeff', email: 'jeff@turing.org', purpose: 'vacation' };
    mockArea = [{ area: "RiNo", details: "/api/v1/areas/590" }];

    fetchDetails.mockImplementation(() => {
      return Promise.resolve([{
        id: 590,
        name: "River North",
        location: "North of Downtown Denver",
        about: "RiNo is a burgeoning area with new bars, restaurants and event spaces popping up all the time. Explore this thriving area of Denver today!",
        region_code: 6356834,
        quick_search: "o5kod9f5cqo0",
        listings: [
        "/api/v1/listings/3",
        "/api/v1/listings/44",
        "/api/v1/listings/221",
        "/api/v1/listings/744",
        "/api/v1/listings/90",
        "/api/v1/listings/310"
        ]
      }])
    })

    getAreas.mockImplementation(() => {
      return Promise.resolve([{
        area: "RiNo",
        details: "/api/v1/areas/590"
      }])
    })
  });

  it('should match the snapshot', () => {
    wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should start off with a default state', () => {
    wrapper = shallow(<App />);
    const expected = {
      isLoggedIn: false,
      areas: '',
      error: '',
      user: { name: '', email: '', purpose: '' },
      isLoading: true,
      listings: [],
      currentListing: '',
      favorites: [],
    };
    expect(wrapper.state()).toEqual(expected);
  });

  it('should retrieve areas when mounting', () => {
    wrapper = shallow(<App />);
    expect(getAreas).toHaveBeenCalled();
  });

  it('should set state with user info when addUser is called', () => {
    wrapper = shallow(<App />);
    wrapper.instance().addUser(mockUser);
    expect(wrapper.state()).toEqual({
      isLoggedIn: true,
      areas: '',
      error: '',
      user: { name: 'Jeff', email: 'jeff@turing.org', purpose: 'vacation' },
      isLoading: true,
      listings: [],
      currentListing: '',
      favorites: [],
    });
  });

  it('should reset user info and favorites in state when logout is called', () => {
    wrapper = shallow(<App />);
    wrapper.instance().addUser(mockUser);
    wrapper.instance().logout();
    expect(wrapper.state()).toEqual({
      isLoggedIn: false,
      areas: '',
      error: '',
      user: { name: '', email: '', purpose: '' },
      isLoading: true,
      listings: [],
      currentListing: '',
      favorites: [],
    });
  });

  it('should set the currentListing in state when setCurrentListing is called', () => {
    wrapper = shallow(<App />);
    wrapper.instance().setCurrentListing(mockListing);
    expect(wrapper.state()).toEqual({
      isLoggedIn: false,
      areas: '',
      error: '',
      user: { name: '', email: '', purpose: '' },
      isLoading: true,
      listings: [],
      currentListing: mockListing,
      favorites: [],
    });
  });

  it('should use area data to retrieve area details when mounting', () => {
    wrapper = shallow(<App />);
    expect(fetchDetails).toHaveBeenCalledWith([{
      area: "RiNo",
      details: "/api/v1/areas/590"
    }]);
  });

  it('should update state with area details', () => {
    const expected = {
        id: 590,
        name: "River North",
        location: "North of Downtown Denver",
        about: "RiNo is a burgeoning area with new bars, restaurants and event spaces popping up all the time. Explore this thriving area of Denver today!",
        region_code: 6356834,
        quick_search: "o5kod9f5cqo0",
        listings: [
        "/api/v1/listings/3",
        "/api/v1/listings/44",
        "/api/v1/listings/221",
        "/api/v1/listings/744",
        "/api/v1/listings/90",
        "/api/v1/listings/310"
        ]
    };
    wrapper = shallow(<App />);
    expect(fetchDetails).toHaveBeenCalledWith(mockArea);
    expect(wrapper.state('areas')).toEqual(expected);
  });

  it('should set state with a new favorite listing when addFavorite is called', () => {
    wrapper = shallow(<App />);
    wrapper.instance().addFavorite(mockListing);
    expect(wrapper.state()).toEqual({
      isLoggedIn: false,
      areas: '',
      error: '',
      user: { name: '', email: '', purpose: '' },
      isLoading: true,
      listings: [],
      currentListing: '',
      favorites: [mockListing],
    });
  });

  it('should remove a favorite listing and set state when removeFavorite is called', () => {
    wrapper = shallow(<App />);
    wrapper.instance().addFavorite(mockListing);
    wrapper.instance().removeFavorite(5);
    expect(wrapper.state()).toEqual({
      isLoggedIn: false,
      areas: '',
      error: '',
      user: { name: '', email: '', purpose: '' },
      isLoading: true,
      listings: [],
      currentListing: '',
      favorites: [],
    });
  });
});
