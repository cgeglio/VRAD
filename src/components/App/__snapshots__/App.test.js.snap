// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<main
  className="app"
>
  <Route
    exact={true}
    path="/"
  >
    <img
      alt="Logo"
      className="logo"
      src="/scoutGreen.png"
    />
    <img
      alt="Rocky Mountain National Park"
      className="background-img"
      src="/rmnp.jpg"
    />
    <Form
      addUser={[Function]}
    />
  </Route>
  <Route
    exact={true}
    path="/areas"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/area/:id/listings"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/area/:id/listings/:id"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/favorites"
    render={[Function]}
  />
</main>
`;
